### ISSUE TITLE
[LightGBM] Regression: penalize negative predictions

### ISSUE THREAD

**Issue by brunocous** [Azure/mmlspark#918]
I have a simple regression task (using a LightGBMRegressor) where I want to penalize negative predictions more than positive ones. Is there a way to achieve this with the default regression LightGBM objectives (see https://lightgbm.readthedocs.io/en/latest/Parameters.html)? If not, is it somehow possible to define (many example for default LightGBM model) and pass a custom regression objective?

Thx in advance!

**Comment by tyler-romero** [comment]
The simplest way might be to duplicate positive examples in your training set so that they naturally carry more influence during training.

**Comment by imatiach-msft** [comment]
"define (many example for default LightGBM model) and pass a custom regression objective"
@brunocous unfortunately, we don't have this currently in the mmlspark wrapper of lightgbm, but it would definitely be a great feature to add at some point.

**Comment by Mihir-Khandelwal** [comment]
It is clear, no bugs, but the color scheme of the website must be suitable to it, it is going other way.

**Comment by imatiach-msft** [comment]
I should update this thread that custom objectives have been added recently, so this should be possible now

**Comment by DefUs3r** [comment]
Hi @imatiach-msft could you let me know if I can work on this?

**Comment by imatiach-msft** [comment]
@DefUs3r sure, what do you want to work on specifically?  Are you thinking of adding an example on how to penalize negative predictions using a custom loss function?

**Comment by Mihir-Khandelwal** [comment]
Hii @brunocous everything is almost fine on your provided link, but it could be more Attractive by using some more UI features, color schemes, and images. Also in your Python code learning, you can give the idea for them starting from the basics.

**Comment by imatiach-msft** [comment]
@Mihir-Khandelwal are you commenting on the right thread?  I think you might have meant to comment elsewhere?

**Comment by Mihir-Khandelwal** [comment]
@imatiach-msft Sorry i am new to this, as an microsoft recruting task, i was assigned a task to comment a code correction on any of the 100+ stars github account.

**Comment by navya002** [comment]
> I have a simple regression task (using a LightGBMRegressor) where I want to penalize negative predictions more than positive ones. Is there a way to achieve this with the default regression LightGBM objectives (see https://lightgbm.readthedocs.io/en/latest/Parameters.html)? If not, is it somehow possible to define (many example for default LightGBM model) and pass a custom regression objective?
> 
> Thx in advance!

I want to ask just for clarity purpose that here ,positive and negatives samples refer to overestimation and underestimation , correct?

**Comment by brunocous** [comment]
> > I have a simple regression task (using a LightGBMRegressor) where I want to penalize negative predictions more than positive ones. Is there a way to achieve this with the default regression LightGBM objectives (see https://lightgbm.readthedocs.io/en/latest/Parameters.html)? If not, is it somehow possible to define (many example for default LightGBM model) and pass a custom regression objective?
> > Thx in advance!
> 
> I want to ask just for clarity purpose that here ,positive and negatives samples refer to overestimation and underestimation , correct?

Correct. In the end, I went with just using the quantile regression objective to penalise overestimations more. That worked enough for my task at hand. Second trick that helped a lot is assigning samples more prone to overestimate a higher sample weight. Not perfect solutions, but it does the trick.

**Comment by navya002** [comment]
Hi ,

Thanks for responding , that  resonates with me as well. Also, while I was
looking at it (the problem) I optimised objective function a bit for better
results since in the 50th percent quantile it turns out to be mae , I
changed it a bit for better results.Please have a look and let me know what
you think (I have submitted the pull request with that function).

Hoping it helps and awaiting your reply.

Thankfully
Navya

**Comment by karnvaibhav11** [comment]
> > I want to ask just for clarity purpose that here ,positive and negatives samples refer to overestimation and underestimation , correct?
> 
> Correct. In the end, I went with just using the quantile regression objective to penalise overestimations more. That worked well enough for my use case. Second trick that helped a lot is assigning samples more prone to overestimate a higher sample weight for training. Not perfect solutions, but it does the trick.

Please assign me this @brunocous @imatiach-msft

**Comment by karnvaibhav11** [comment]
Hi,
I have found a way to penalize the negative prediction in Regression by using the concepts of Adaboost penalization of wrong Classification Penalization. Basically what it does is, it updates the weight of the negative residual and increase its training example for the next model. This ensures that if the value is overestimated it penalizes it more.  It's a completely new concept which I researched and the work is still in progress.  
Please find the attached Docx. which has the steps that I followed
and pdf for the code. Note that I haven't created the intervals due to the lack of time and to meet my Microsoft Project deadline. But I'll work on it later on.
Also, I am unable to create a pull request so I am sharing these files over here. Will love to hear your feedback. Thanks.

[Penalizing the Negative Prediction in a Regression Problem.docx](https://github.com/microsoft/SynapseML/files/7083441/Penalizing.the.Negative.Prediction.in.a.Regression.Problem.docx)
[M2LinearRegression.pdf](https://github.com/microsoft/SynapseML/files/7083523/M2LinearRegression.pdf)

**Comment by ParulSrivastava** [comment]
[Penalizing.the.Negative.Prediction.in.a.Regression.Problem.docx](https://github.com/microsoft/SynapseML/files/7655763/Penalizing.the.Negative.Prediction.in.a.Regression.Problem.docx)


I have Created a pipeline using Azure ML for Regression Model

**Comment by ranadeepsingh** [comment]
Thanks for the thoughtful discussion here. There are now two supported paths for this use case:

- As the original reporter noted, quantile regression and sample weighting can penalize overestimation differently: https://github.com/microsoft/SynapseML/issues/918#issuecomment-908303142
- SynapseML LightGBM now exposes custom objectives through the `fobj` parameter: https://github.com/microsoft/SynapseML/blob/v1.1.3/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/params/LightGBMParams.scala

Since the question has workable built-in and custom-objective answers, I am closing it as resolved. A focused documentation PR with a custom-loss example would still be welcome.
